apps/arena/lib/arena/game/skill.ex

```diff
@@ -232,15 +257,101 @@ defmodule Arena.Game.Skill do
 
       Process.send_after(self(), {:remove_effect, player.id, last_id}, effect.duration_ms)
 
-      effects =
-        get_in(game_state, [:players, player.id, :aditional_info, :effects])
-        |> Map.put(last_id, effect)
-
-      put_in(game_state, [:players, player.id, :aditional_info, :effects], effects)
+      put_in(
+        game_state,
+        [:players, player.id, :aditional_info, :effects, last_id],
+        Map.put(effect, :id, last_id)
+      )
       |> put_in([:last_id], last_id)
     end)
   end
 
+  def apply_effect_mechanic(players, game_state) do
+    updated_players =
+      Map.filter(players, fn {_, player} -> Player.alive?(player) end)
+      |> Map.new(fn {_player_id, player} ->
+        player =
+          Enum.reduce(player.aditional_info.effects, player, fn {_effect_id, effect}, player ->
+            apply_effect_mechanic(player, effect, game_state)
+          end)
+
+        {player.id, player}
+      end)
+
+    Map.merge(players, updated_players)
+  end
+
+  def apply_effect_mechanic(player, effect, game_state) do
+    now = DateTime.utc_now() |> DateTime.to_unix(:millisecond)
+
+    Enum.reduce(effect.effect_mechanics, player, fn {mechanic_name, mechanic_params} = mechanic,
+                                                    player ->
+      should_re_apply? =
+        now - Map.get(mechanic_params, :last_application_time, 0) >=
+          mechanic_params.effect_delay_ms
+
+      if should_re_apply? do
+        do_effect_mechanics(game_state, player, effect, mechanic)
+        |> put_in(
+          [
+            :aditional_info,
+            :effects,
+            effect.id,
+            :effect_mechanics,
+            mechanic_name,
+            :last_application_time
+          ],
+          now
+        )
+      else
+        player
+      end
+    end)
+  end
+
+  defp do_effect_mechanics(game_state, player, effect, {:pull, pull_params}) do
+    Map.get(game_state.pools, effect.owner_id)
+    |> case do
+      nil ->
+        player
+
+      pool ->
+        if pool.position != player.position do
```

Comment:
This `if` should be another branch of the `case`

For example

```elixir
player_position = player.position
case ... do
  nil -> player
  %{pool: position} when position == player_position -> player
  pool -> 
    ...
end
```

Reply:
done! it's much better